### PR TITLE
[WIP] Fixed Timestep Interpolation (3D) - version four

### PR DESCRIPTION
**Pull request by lawnjelly**
Adds fixed timestep interpolation to the visual server.
Loosely based around godotengine/godot-proposals#2753

This is based on the latest suggestions from reduz. His latest suggestion is behaviour dependent on the node type, hopefully I have this correct from our last discussions:

### Physics nodes
Interpolate by default (teleport when you call `set_transform`, I'm assuming)

### Cameras
Interpolate by default, presumably do not teleport when you call `set_transform` (as that would be needed to move the camera). No mechanism to call teleport yet.

### Other Spatials
Do not interpolate.

### New node type
reduz suggested rather than allowing Spatials to turn off interpolation, we should add a new node type with the sole purpose of turning on and off interpolation. (?) Not yet implemented.

## Notes
* This isn't a full implementation, I don't think it is worth me doing any further work until a decision is made on which approach to take.
* I would personally use #52846, and imo is far better than any of the other suggestions, it is just the simple standard implementation of fixed timestep, each node can turn on and off interpolation, and call `teleport()`.
* The implementation is this PR is highly convoluted and hard to follow, with the user having to understand that for some reason we have different behaviour for different node types. This makes no sense to me, both from a user and engine perspective.

### CODE REVIEW

File: servers/visual/visual_server_scene.cpp [hunk 1]
```diff
@@ -675,30 +704,238 @@ void VisualServerScene::instance_set_layer_mask(RID p_instance, uint32_t p_mask)
...
+}
+
 void VisualServerScene::instance_set_transform(RID p_instance, const Transform &p_transform) {
 	Instance *instance = instance_owner.get(p_instance);
 	ERR_FAIL_COND(!instance);
 
-	if (instance->transform == p_transform) {
-		return; //must be checked to avoid worst evil
-	}
+	if (!Engine::get_singleton()->is_physics_interpolation_enabled() || !instance->interpolated) {
```
Comment (jordo) [inline servers/visual/visual_server_scene.cpp:722]: I commented the same in rocket chat, but would love to see any settings for interpolation not be engine/global.   As reduz suggested perhaps a property on scene tree, and perhaps something gets assigned when a node is entered in that tree.

File: servers/visual/visual_server_scene.cpp [hunk 2]
```diff
@@ -675,30 +704,238 @@ void VisualServerScene::instance_set_layer_mask(RID p_instance, uint32_t p_mask)
...
+	}
+
+	// if the camera was on any of the lists, remove
+	camera_transform_update_list_curr->erase_multiple_unordered(p_rid);
+	camera_transform_update_list_prev->erase_multiple_unordered(p_rid);
+	camera_teleport_list.erase_multiple_unordered(p_rid);
 }
+
+void VisualServerScene::InterpolationData::notify_free_instance(RID p_rid) {
+	if (!Engine::get_singleton()->is_physics_interpolation_enabled()) {
```
Comment (jordo) [inline servers/visual/visual_server_scene.cpp:810]: Again, prev comment regarding engine setting global

File: servers/visual/visual_server_scene.cpp [hunk 3]
```diff
@@ -4059,6 +4307,7 @@ bool VisualServerScene::free(RID p_rid) {
 
 		instance_owner.free(p_rid);
 		memdelete(instance);
+		_interpolation_data.notify_free_instance(p_rid);
```
Comment (jordo) [inline servers/visual/visual_server_scene.cpp:4310]: these looks safe to notify after freeing the instance, but what if some one changes `notify_free_instance` in some unexpected fashion later on?  Should not `memdelete(instance);` be the last operation here?

**Review by jordo — COMMENTED** [review]
Thanks for your work @lawnjelly, i know this one is a difficult one.  I'll try and spend a little time on the other PRs as well.

**Comment by lawnjelly** [comment]
I'm just having a look at whether we can simply store the interpolation lists in the `Scenario`. The scenario seems to be created from the scene tree so this may be a solution to this global nature for multiple scene trees.

The only tricky bit is that the instances and cameras need quick access to the interpolation lists, for the `set_transform` commands and also the `free` commands. The instances store a Scenario pointer but the Cameras do not appear to, but that could be potentially be added (I don't know if there are any situations a Camera is used without a Scenario?).

Then you could turn on and off the global interpolation setting per scene tree (or scenario, if you weren't using a scene tree), which would hopefully deal with these use cases.

**Comment by lawnjelly** [comment]
Closing (at least temporarily) in favour of #52846.
